plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.test.ts

```diff
@@ -23,33 +23,187 @@ import {
   DefaultProcessingDatabase,
 } from './DefaultProcessingDatabase';
 
-import { Entity, EntityRelationSpec } from '@backstage/catalog-model';
+import { Entity, stringifyEntityRef } from '@backstage/catalog-model';
 import * as uuid from 'uuid';
 import { getVoidLogger } from '@backstage/backend-common';
+import { JsonObject } from '@backstage/config';
 
 describe('Default Processing Database', () => {
   let db: Knex;
   let processingDatabase: DefaultProcessingDatabase;
   const logger = getVoidLogger();
 
+  const insertRefRow = async (ref: DbRefreshStateReferencesRow) => {
+    return db<DbRefreshStateReferencesRow>('refresh_state_references').insert(
+      ref,
+    );
+  };
+
+  const insertRefreshStateRow = async (ref: DbRefreshStateRow) => {
+    await db<DbRefreshStateRow>('refresh_state').insert(ref);
+  };
+
   beforeEach(async () => {
     db = await DatabaseManager.createTestDatabaseConnection();
     await DatabaseManager.createDatabase(db);
 
     processingDatabase = new DefaultProcessingDatabase(db, logger);
   });
 
-  describe('replaceUnprocessedEntities', () => {
-    const insertRefRow = async (ref: DbRefreshStateReferencesRow) => {
-      return db<DbRefreshStateReferencesRow>('refresh_state_references').insert(
-        ref,
+  describe('updateProcessedEntity', () => {
+    let id: string;
+    let processedEntity: Entity;
+
+    beforeEach(() => {
+      id = uuid.v4();
+      processedEntity = {
+        apiVersion: '1',
+        kind: 'Location',
+        metadata: {
+          name: 'fakelocation',
+        },
+        spec: {
+          type: 'url',
+          target: 'somethingelse',
+        },
+      };
+    });
+
+    it('fails when there is no processing state for the entity', async () => {
+      await processingDatabase.transaction(async tx => {
+        await expect(() =>
+          processingDatabase.updateProcessedEntity(tx, {
+            id,
+            processedEntity,
+            state: new Map<string, JsonObject>(),
+            relations: [],
+            deferredEntities: [],
+          }),
+        ).rejects.toThrow(`Processing state not found for ${id}`);
+      });
+    });
+
+    it('updates the refresh state entry with the cache, processed entity and errors', async () => {
+      await insertRefreshStateRow({
+        entity_id: id,
+        entity_ref: 'location:default/fakelocation',
+        unprocessed_entity: '{}',
+        processed_entity: '{}',
+        errors: '',
+        next_update_at: 'now()',
+        last_discovery_at: 'now()',
+      });
+
+      const state = new Map<string, JsonObject>();
+      state.set('hello', { t: 'something' });
+
+      await processingDatabase.transaction(tx =>
+        processingDatabase.updateProcessedEntity(tx, {
+          id,
+          processedEntity,
+          state,
+          relations: [],
+          deferredEntities: [],
+          errors: 'something broke',
```

Comment:
this is defined to be a stringified `Error[]` (so an array); of course the test doesn't care but if you wanted to choose a value here then maybe `'[]'` would be clearest